vita3k/util/include/util/fs.h

```diff
@@ -19,9 +19,156 @@
 
 #include <boost/filesystem.hpp>
 #include <boost/filesystem/fstream.hpp>
+#include <fmt/format.h>
+#include <fmt/std.h>
 
 namespace fs = boost::filesystem;
 
+#ifdef _WIN32
+#define FOPEN(filename, params) _wfopen(filename, L##params)
+#else
+#define FOPEN(filename, params) fopen(filename, params)
+#endif
+
+namespace fs_utils {
+
+inline const auto dir_sep = fs::path::string_type{ fs::path::preferred_separator };
+
+/**
+ * \brief  Construct a file name (optionally with an extension) to be placed in a Vita3K directory.
+ * \param  base_path   The main output path for the file.
+ * \param  folder_path The sub-directory/sub-directories to output to.
+ * \param  file_name   The name of the file.
+ * \param  extension   The extension of the file (optional)
+ * \return A complete Boost.Filesystem file path normalized.
+ */
+fs::path construct_file_name(const fs::path &base_path, const fs::path &folder_path, const fs::path &file_name, const fs::path &extension = "");
+
+/**
+ * \brief Convert a Boost.Filesystem path to a UTF-8 string.
+ * \param path path to convert
+ * \return UTF-8 string
+ */
+std::string path_to_utf8(const fs::path &path);
+
+/**
+ * \brief Convert a UTF-8 string to a Boost.Filesystem path.
+ * \param str UTF-8 string to convert
+ * \return Boost.Filesystem path
+ */
+fs::path utf8_to_path(const std::string &str);
+
+/**
+ * \brief Concatenate two Boost.Filesystem paths. Usable to add extension to path.
+ * \param path1 First path
+ * \param path2 Second path
+ * \return First path + second path
+ */
+fs::path path_concat(const fs::path &path1, const fs::path &path2);
+
+/**
+ * \brief Check if a directory exists and create the directory and all its parents if they don't exist.
+ * \param path Directory to check
+ * \return True if the directory exists or was created successfully, false otherwise.
+ */
+bool force_directories(const fs::path &path);
```

Comment:
fs::create_directories already does exactly this (the call to fs::exists before is useless)

Reply:
I've seen such code (`if (!fs::exists(dir)) create_directories(dir);`) If it's ok to just call `create_directories(dir);` it's great

Reply:
On the documentation of `create_directories(dir);` for both std and boost, it is written that this function has no effect if the directory exists. The only case where force_directories would behave differently compared to just calling create_directories is if dir already exists but as a file, create_directories would throw an exception while force_directories will do nothing, I am fine with the exception as this defintely wouldn't be expected.

Reply:
Great. Thank you. I replaced force_directories to create_directories
(Also I checked zip unpacking and now it's ready to review)